model/task_queue_service_test.go

```diff
@@ -1200,6 +1201,155 @@ func (s *taskDAGDispatchServiceSuite) TestNextTaskForDefaultTaskSpec() {
 	// .....
 }
 
+func (s *taskDAGDispatchServiceSuite) TestFindNextTaskThreadSafe() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	s.Require().NoError(db.ClearCollections(task.Collection))
+	items := []TaskQueueItem{}
+	for i := 0; i < 100; i++ {
+		items = append(items, TaskQueueItem{
+			Id:            fmt.Sprintf("%d", i),
+			BuildVariant:  "variant_1",
+			Version:       "version_1",
+			Project:       "project_1",
+			GroupMaxHosts: 0,
+		})
+		t := task.Task{
+			Id:                fmt.Sprintf("%d", i),
+			BuildVariant:      "variant_1",
+			Version:           "version_1",
+			Project:           "project_1",
+			TaskGroupMaxHosts: 0,
+		}
+		s.Require().NoError(t.Insert())
+	}
+
+	s.taskQueue.Queue = items
+	service, err := newDistroTaskDAGDispatchService(s.taskQueue, time.Minute)
+	s.NoError(err)
+	spec := TaskSpec{
+		BuildVariant: "variant_1",
+		Version:      "version_1",
+		Project:      "project_1",
+	}
+
+	numGoroutines := 100
+
+	var wg sync.WaitGroup
+	for i := 0; i < numGoroutines; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			service.FindNextTask(ctx, spec, utility.ZeroTime)
+		}()
+	}
+	wg.Wait()
+
+	dispatchedCount := 0
+	for _, item := range service.nodeItemMap {
+		if item.IsDispatched {
+			dispatchedCount++
+		}
+	}
+	s.Equal(dispatchedCount, 100)
+}
+
+func (s *taskDAGDispatchServiceSuite) TestFindNextTaskGroupTaskThreadSafe() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	s.Require().NoError(db.ClearCollections(task.Collection))
+	items := []TaskQueueItem{}
+	for i := 0; i < 20; i++ {
+		id := fmt.Sprintf("%d", i)
+		items = append(items, TaskQueueItem{
+			Id:            id,
+			Group:         "group_1",
+			BuildVariant:  "variant_1",
+			Version:       "version_1",
+			Project:       "project_1",
+			GroupMaxHosts: 1,
+		})
+		t := task.Task{
+			Id:                id,
+			TaskGroup:         "group_1",
+			BuildVariant:      "variant_1",
+			Version:           "version_1",
+			TaskGroupMaxHosts: 1,
+			Project:           "project_1",
+			StartTime:         utility.ZeroTime,
+			FinishTime:        utility.ZeroTime,
+		}
+		s.Require().NoError(t.Insert())
+
+		s.taskQueue = TaskQueue{
+			Distro: "distro_1",
+			Queue:  items,
+		}
+	}
+
+	s.taskQueue.Queue = items
+	service, err := newDistroTaskDAGDispatchService(s.taskQueue, time.Minute)
+	s.NoError(err)
+
+	// Populating the Group field on the task spec indicates the host just ran a task group.
+	spec := TaskSpec{
+		Group:        "group_1",
+		BuildVariant: "variant_1",
+		Version:      "version_1",
+		Project:      "project_1",
+	}
+
+	numGoroutines := 20
+	var wg sync.WaitGroup
+	wg.Add(numGoroutines)
+	for i := 0; i < numGoroutines; i++ {
+		go func() {
+			defer wg.Done()
```

Comment:
It may help to add a channel to block the goroutines from running (like [this](https://github.com/evergreen-ci/evergreen/blob/003a08f1f323d0ea0c90e944a2fbb5c5c799ee70/cloud/parameterstore/fakeparameter/parameter_manager_test.go#L362) before they're all initialized. Since the goroutines are initialized serially, some goroutines might already be done by the time others start, which reduces the amount of concurrency. If they're not running concurrently, it reduces the likelihood that a race/some kind of dispatch contention bug would manifest if there is one.

Same with the other goroutines.